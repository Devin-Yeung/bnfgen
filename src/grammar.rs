use rand::distributions::Distribution;
use rand::distributions::WeightedIndex;
use rand::Rng;
use std::collections::HashMap;

use crate::{lexer, parser};

#[repr(transparent)]
#[derive(Debug)]
pub struct RawGrammar {
    pub(crate) rules: Vec<Rule>,
}

impl RawGrammar {
    pub fn parse<S: AsRef<str>>(input: S) -> RawGrammar {
        let lexer = lexer::Lexer::new(input.as_ref());
        let parser = parser::RawGrammarParser::new();
        match parser.parse(lexer) {
            Ok(grammar) => grammar,
            Err(e) => match e {
                lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                    panic!("Unrecognized token: {:?}, expected: {:?}", token, expected)
                }
                lalrpop_util::ParseError::ExtraToken { token } => {
                    panic!("Extra token: {:?}", token)
                }
                lalrpop_util::ParseError::User { error } => panic!("User error: {:?}", error),
                lalrpop_util::ParseError::InvalidToken { location } => {
                    panic!("Invalid token: {:?}", location)
                }
                lalrpop_util::ParseError::UnrecognizedEof { location, expected } => {
                    panic!("Unrecognized EOF: {:?}, expected: {:?}", location, expected)
                }
            },
        } // TODO: Error Handling
    }
}

pub struct CheckedGrammar {
    pub(crate) rules: HashMap<String, WeightedProduction>,
}

impl CheckedGrammar {
    fn reduce(&self, symbol: SymbolKind) -> (Option<String>, Vec<SymbolKind>) {
        match symbol {
            SymbolKind::Terminal(s) => (Some(s), vec![]),
            SymbolKind::NonTerminal(s) => {
                let syms = self.rules.get(&s).unwrap().choose(&mut rand::thread_rng());
                (None, syms)
            }
            SymbolKind::Repeat { symbol, min, max } => {
                todo!()
            }
        }
    }
}

#[derive(Debug)]
pub struct Rule {
    pub(crate) name: String,
    pub(crate) production: WeightedProduction,
}

#[derive(Debug)]
#[repr(transparent)]
pub struct WeightedProduction {
    pub(crate) production: Vec<Alternative>,
}

impl WeightedProduction {
    pub fn choose<R: Rng>(&self, rng: &mut R) -> Vec<SymbolKind> {
        let dist = WeightedIndex::new(self.production.iter().map(|a| a.weight)).unwrap();
        let idx = dist.sample(rng);
        self.production[idx]
            .symbols
            .iter()
            .map(|s| s.kind.clone())
            .collect()
    }
}

#[derive(Debug)]
pub struct Alternative {
    pub(crate) weight: usize,
    pub(crate) symbols: Vec<Symbol>,
}

#[derive(Debug, Clone)]
pub enum SymbolKind {
    Terminal(String),
    NonTerminal(String),
    Repeat {
        symbol: Box<SymbolKind>,
        min: usize,
        max: Option<usize>,
    },
}

#[derive(Debug)]
pub struct Symbol {
    pub(crate) kind: SymbolKind,
}

#[cfg(test)]
mod test {
    use super::RawGrammar;

    #[test]
    fn brainfuck() {
        let text = include_str!("../examples/brainfuck.bnfgen");
        let grammar = RawGrammar::parse(text);
        insta::assert_debug_snapshot!(grammar);
    }
}
