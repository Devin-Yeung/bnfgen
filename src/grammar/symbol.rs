use crate::regex::Regex;
use crate::span::Span;
use std::hash::Hash;
use std::rc::Rc;

pub type Terminal = Rc<String>;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct NonTerminal {
    pub(crate) name: Rc<String>,
    pub(crate) ty: Ty,
}

impl Hash for NonTerminal {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
        self.ty.hash(state);
    }
}

impl NonTerminal {
    pub fn untyped<S: Into<String>>(name: S) -> Self {
        NonTerminal {
            name: Rc::new(name.into()),
            ty: Ty::Untyped,
        }
    }

    pub fn typed<S: Into<String>>(name: S, ty: Ty) -> Self {
        NonTerminal {
            name: Rc::new(name.into()),
            ty,
        }
    }

    pub fn as_str(&self) -> &str {
        self.name.as_str()
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum Ty {
    Untyped,
    Typed(Rc<String>),
}

impl From<String> for Ty {
    fn from(s: String) -> Self {
        Ty::Typed(Rc::new(s))
    }
}

impl Hash for Ty {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        match self {
            Ty::Untyped => { /* do nothing */ }
            Ty::Typed(s) => s.hash(state),
        }
    }
}

impl Ty {
    pub fn untyped() -> Self {
        Ty::Untyped
    }

    pub fn typed<S: Into<String>>(s: S) -> Self {
        Ty::Typed(Rc::new(s.into()))
    }
}

impl Clone for Ty {
    fn clone(&self) -> Self {
        match self {
            Ty::Untyped => Ty::Untyped,
            Ty::Typed(s) => Ty::Typed(s.clone()),
        }
    }
}

impl Ty {
    pub fn ty(&self) -> Option<&str> {
        match self {
            Ty::Untyped => None,
            Ty::Typed(s) => Some(s.as_str()),
        }
    }
}

#[derive(Debug, Clone)]
pub enum SymbolKind {
    Terminal(Terminal),
    NonTerminal(NonTerminal),
    Regex(Rc<Regex>),
}

impl Hash for SymbolKind {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        match self {
            SymbolKind::Terminal(s) => s.hash(state),
            SymbolKind::NonTerminal(s) => s.hash(state),
            SymbolKind::Regex(s) => s.hash(state),
        }
    }
}

impl SymbolKind {
    pub fn non_re_terminal(&self) -> Option<&str> {
        match self {
            SymbolKind::Terminal(s) => Some(s.as_str()),
            _ => None,
        }
    }

    pub fn is_terminal(&self) -> bool {
        matches!(self, SymbolKind::Terminal(_) | SymbolKind::Regex(_))
    }

    // get the non-terminal symbol if it is a non-terminal symbol, else none
    pub fn non_terminal(&self) -> Option<&str> {
        match self {
            SymbolKind::NonTerminal(s) => Some(s.as_str()),
            _ => None,
        }
    }
}

#[derive(Debug)]
pub struct Symbol {
    pub(crate) kind: SymbolKind,
    pub(crate) span: Span,
}

impl Hash for Symbol {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.kind.hash(state);
    }
}

impl Symbol {
    pub fn is_terminal(&self) -> bool {
        self.kind.is_terminal()
    }

    pub fn non_terminal(&self) -> Option<&str> {
        self.kind.non_terminal()
    }
}
