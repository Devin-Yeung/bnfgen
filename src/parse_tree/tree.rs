use std::fmt;
use std::fmt::Debug;

pub enum ParseTree<T> {
    Leaf(T),
    Branch {
        name: String,
        children: Vec<ParseTree<T>>,
    },
}

impl<T> ParseTree<T> {
    pub(crate) fn branch(name: String) -> ParseTree<T> {
        ParseTree::Branch {
            name,
            children: Vec::new(),
        }
    }

    pub(crate) fn children_len(&self) -> usize {
        match self {
            ParseTree::Leaf(_) => 1, // TODO: 0 or 1 ?
            ParseTree::Branch { children, .. } => children.len(),
        }
    }
}

// derive debug if T: Debug
impl<T: Debug> Debug for ParseTree<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ParseTree::Leaf(value) => f.debug_tuple("Leaf").field(value).finish(),
            ParseTree::Branch { name, children } => f
                .debug_struct("Branch")
                .field("name", name)
                .field("children", children)
                .finish(),
        }
    }
}
