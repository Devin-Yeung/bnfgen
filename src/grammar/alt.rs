use crate::grammar::symbol::Symbol;
use crate::span::Span;
use std::hash::{DefaultHasher, Hash, Hasher};

#[derive(Debug)]
pub enum Limit {
    /// can be invoked any number of times
    Unlimited,
    Limited {
        /// should be invoked at least `min` times (inclusive)
        min: usize,
        /// should be invoked at most `max` times (inclusive)
        max: usize,
    },
}

#[derive(Debug)]
pub struct Alternative {
    pub(crate) span: Span,
    pub(crate) weight: usize,
    pub(crate) invoke_limit: Limit,
    pub(crate) symbols: Vec<Symbol>,
}

impl Hash for Alternative {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.symbols.hash(state);
    }
}

pub type AltId = u64;

impl Alternative {
    /// returns the non-regex terminals in this alternative
    pub(crate) fn non_re_terminals(&self) -> Vec<&str> {
        self.symbols
            .iter()
            .filter_map(|s| s.kind.non_re_terminal())
            .collect()
    }

    /// return the unique id of this alternative
    pub(crate) fn id(&self) -> AltId {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);
        hasher.finish()
    }
}
