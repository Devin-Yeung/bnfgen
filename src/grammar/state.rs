use crate::grammar::alt::AltId;
use crate::grammar::symbol::Ty;
use rand::Rng;
use std::collections::HashMap;
use std::rc::Rc;

pub struct State<R: Rng> {
    rng: R,
    /// tracking the number of times an alternative has been selected
    /// Notes: only those with invoke limits are tracked
    pub(crate) tracking: HashMap<AltId, usize>,
    /// tracking the declared variable and its type
    pub(crate) vars: HashMap<String, Ty>,
    /// tracking the post declared variable
    pub(crate) waiting_to_declared: HashMap<String, Ty>,
}

impl<R: Rng> State<R> {
    pub fn new(rng: R) -> Self {
        Self {
            rng,
            tracking: HashMap::new(),
            vars: HashMap::new(),
            waiting_to_declared: HashMap::new(),
        }
    }

    pub fn rng(&mut self) -> &mut R {
        &mut self.rng
    }

    pub fn track(&mut self, id: AltId) {
        let count = self.tracking.entry(id).or_insert(0);
        *count += 1;
    }

    pub fn count(&self, id: AltId) -> usize {
        *self.tracking.get(&id).unwrap_or(&0)
    }
}
