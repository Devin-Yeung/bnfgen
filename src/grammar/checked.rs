use crate::grammar::production::WeightedProduction;
use crate::grammar::state::State;
use crate::grammar::symbol::{NonTerminal, SymbolKind};
use rand::Rng;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Debug)]
pub struct CheckedGrammar {
    pub(crate) rules: HashMap<NonTerminal, WeightedProduction>,
}

pub enum ReduceOutput {
    Terminal(Rc<String>),
    NonTerminal {
        name: Rc<String>,
        syms: Vec<SymbolKind>,
    },
}

impl CheckedGrammar {
    /// '+' --reduce--> '+'
    ///
    /// E   --reduce--> E, remaining: ['+', E]
    /// if E -> E '+' E
    pub(crate) fn reduce<R: Rng>(&self, symbol: SymbolKind, state: &mut State<R>) -> ReduceOutput {
        match symbol {
            SymbolKind::Terminal(s) => ReduceOutput::Terminal(s),
            SymbolKind::NonTerminal(s) => {
                let syms = self
                    .rules
                    .get(&s)
                    .unwrap_or_else(|| panic!("Fail to find rule of {:?}", s))
                    .choose_by_state(state);

                ReduceOutput::NonTerminal { name: s.name, syms }
            }
            SymbolKind::Regex(re) => {
                let terminals = self
                    .rules
                    .values()
                    .flat_map(|r| r.non_re_terminals())
                    .collect::<Vec<_>>();
                let s = re.generate(state.rng(), terminals.as_slice());
                ReduceOutput::Terminal(Rc::new(s))
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::grammar::raw::RawGrammar;

    #[test]
    fn it_can_merge() {
        let text = r#"
            <E> ::= <E: "int"> "+" <E: "int"> ;
            <E> ::= <E: "str"> "+" <E: "str"> ;
            <E: "str"> ::= <E: "str"> "+" <E: "str"> ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap();
        insta::assert_debug_snapshot!(grammar.to_checked().unwrap());
    }
}
