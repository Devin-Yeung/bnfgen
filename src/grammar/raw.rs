use crate::error::Error;
use crate::grammar::alt::Limit;
use crate::grammar::checked::CheckedGrammar;
use crate::grammar::graph::GrammarGraph;
use crate::grammar::rule::Rule;
use crate::grammar::symbol::SymbolKind;
use crate::lexer;
use crate::span::Span;
use crate::utils::convert_parse_error;
use petgraph::graph::{DiGraph, NodeIndex};
use std::collections::{HashMap, HashSet};

#[repr(transparent)]
#[derive(Debug)]
pub struct RawGrammar {
    pub(crate) rules: Vec<Rule>,
}

impl RawGrammar {
    pub fn parse<S: AsRef<str>>(input: S) -> crate::error::Result<RawGrammar> {
        let lexer = lexer::Lexer::new(input.as_ref());
        let parser = crate::parser::RawGrammarParser::new();
        parser.parse(lexer).map_err(convert_parse_error)
    }

    pub fn to_checked(self) -> crate::error::Result<CheckedGrammar> {
        self.check_undefined()?.check_duplicate()?.check_repeats()?;

        let mut rules = HashMap::new();
        for rule in self.rules {
            rules.insert(rule.name, rule.production);
        }

        Ok(CheckedGrammar { rules })
    }

    pub fn graph(&self) -> GrammarGraph<'_> {
        let mut graph = DiGraph::<String, ()>::new();
        let nodes: HashMap<String, NodeIndex> = self
            .rules
            .iter()
            .map(|rule| {
                let entry = (rule.name.clone(), graph.add_node(rule.name.clone()));
                entry
            })
            .collect();
        // setup the graph
        for rule in &self.rules {
            for sym in rule.rhs().iter().flat_map(|a| a.symbols.iter()) {
                match sym.non_terminal() {
                    Some(name) => {
                        graph.add_edge(nodes[&rule.name], nodes[name], ());
                    }
                    None => { /* do nothing */ }
                }
            }
        }
        GrammarGraph {
            rules: &self.rules,
            graph,
            nodes,
        }
    }

    pub fn check_duplicate(&self) -> crate::error::Result<&Self> {
        let mut seen: HashMap<String, Span> = HashMap::new();
        for rule in &self.rules {
            if let Some(prev) = seen.get(&rule.name) {
                return Err(Error::DuplicatedRules {
                    span: rule.span,
                    prev: *prev,
                });
            }
            seen.insert(rule.name.clone(), rule.span);
        }
        Ok(self)
    }

    pub fn check_repeats(&self) -> crate::error::Result<&Self> {
        for rule in &self.rules {
            for alt in rule.rhs() {
                if let Limit::Limited { min, max } = alt.invoke_limit {
                    if min > max {
                        return Err(Error::InvalidRepeatRange { span: alt.span });
                    }
                }
            }
        }
        Ok(self)
    }

    pub fn check_undefined(&self) -> crate::error::Result<&Self> {
        let defined: HashSet<String> =
            HashSet::from_iter(self.rules.iter().map(|r| r.name.clone()));
        for rule in &self.rules {
            for sym in rule.rhs().iter().flat_map(|a| a.symbols.iter()) {
                match &sym.kind {
                    SymbolKind::NonTerminal(s) => {
                        if !defined.contains(s.as_str()) {
                            return Err(Error::UndefinedNonTerminal { span: sym.span });
                        }
                    }
                    _ => { /* do nothing */ }
                }
            }
        }
        Ok(self)
    }
}
