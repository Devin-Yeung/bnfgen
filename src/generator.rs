use crate::grammar::checked::{CheckedGrammar, ReduceOutput};
use crate::grammar::state::State;
use crate::grammar::symbol::SymbolKind::Terminal;
use crate::grammar::symbol::{NonTerminal, SymbolKind};
use crate::parse_tree::tree::ParseTree;
use rand::Rng;

#[derive(typed_builder::TypedBuilder)]
pub struct Generator {
    pub grammar: CheckedGrammar,
}

impl Generator {
    pub fn generate<R: Rng, S: Into<String>>(&self, start: S, rng: &mut R) -> String {
        let mut buf = Vec::new();
        let mut state = State::new(rng);

        let start = SymbolKind::NonTerminal(NonTerminal::untyped(start));
        let mut stack = vec![start];

        while !stack.is_empty() {
            // pop out the first symbol
            match self.grammar.reduce(stack.remove(0), &mut state) {
                ReduceOutput::Terminal(s) => {
                    buf.push(s);
                }
                ReduceOutput::NonTerminal { mut syms, .. } => {
                    // syms :: stack
                    syms.extend(stack);
                    stack = syms;
                }
            }
        }

        buf.iter().map(|s| s.as_str()).collect::<Vec<_>>().join(" ")
    }
}

pub struct TreeGenerator {
    pub grammar: CheckedGrammar,
}

impl TreeGenerator {
    pub fn generate<R: Rng, S: Into<String>>(
        &self,
        start: S,
        rng: &mut R,
    ) -> ParseTree<SymbolKind> {
        let start = SymbolKind::NonTerminal(NonTerminal::untyped(start));
        let mut state = State::new(rng);
        self.generate_tree(start, &mut state)
    }

    fn generate_tree<R: Rng>(
        &self,
        symbol: SymbolKind,
        state: &mut State<R>,
    ) -> ParseTree<SymbolKind> {
        match self.grammar.reduce(symbol, state) {
            ReduceOutput::Terminal(s) => ParseTree::leaf(Terminal(s)),
            ReduceOutput::NonTerminal { name, syms } => {
                let subtrees = syms
                    .into_iter()
                    .map(|sym| self.generate_tree(sym, state))
                    .collect::<Vec<_>>();
                ParseTree::branch(name.to_string(), subtrees)
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::generator::{Generator, TreeGenerator};
    use crate::grammar::raw::RawGrammar;
    use rand::SeedableRng;

    #[test]
    fn repeat_works() {
        let text = r#"
            <S> ::= <E> | <S> <E> {100};
            <E> ::= "a" {1, 10} | "b" {2, } | "c" {3} | "fallback" ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap().to_checked().unwrap();
        let gen = Generator::builder().grammar(grammar).build();
        let out = gen.generate("S", &mut rand::thread_rng());
        assert!(out.split(" ").count() >= 100);
    }

    #[test]
    fn test_tree_generator() {
        let text = r#"
            <S> ::= <E> | <S> <E> {10};
            <E> ::= "a" {1, 10} | "b" {2, } | "c" {3} | "fallback" ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap().to_checked().unwrap();
        let tree_gen = TreeGenerator { grammar };
        let mut seeded_rng = rand::rngs::StdRng::seed_from_u64(42);
        let tree = tree_gen.generate("S", &mut seeded_rng);
        insta::assert_debug_snapshot!(&tree);
    }

    #[test]
    fn test_typed_generator() {
        let text = r#"
            <S> ::= <Expr> | <S> "\n" <Expr> {10, 20};

            <Expr> ::= <E> ;

            <E: "int">  ::= "1" | "2" | "3"
                            | <E: "int"> "+" <E: "int"> {3, } ;

            <E: "bool"> ::= "true" | "false"
                            | <E: "bool"> "&" <E: "bool"> {3, } ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap().to_checked().unwrap();
        let gen = Generator { grammar };
        let mut seeded_rng = rand::rngs::StdRng::seed_from_u64(42);
        insta::assert_snapshot!(gen.generate("S", &mut seeded_rng));
    }

    #[test]
    fn test_typed_set_algebra() {
        let text = include_str!("../examples/set-algebra-typed.bnfgen");
        let grammar = RawGrammar::parse(text).unwrap().to_checked().unwrap();
        let gen = Generator { grammar };
        let mut seeded_rng = rand::rngs::StdRng::seed_from_u64(42);
        insta::assert_snapshot!(gen.generate("Program", &mut seeded_rng));
    }
}
