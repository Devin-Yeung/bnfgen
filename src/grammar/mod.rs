pub mod alt;
pub mod checked;
pub mod graph;
pub mod production;
pub mod raw;
pub mod rule;
pub mod state;
pub mod symbol;

#[cfg(test)]
mod test {
    use crate::grammar::raw::RawGrammar;
    use crate::report::{Reporter, Style};
    use miette::{Diagnostic, Report};
    use std::sync::Arc;

    fn report_with_unnamed_source<T: Diagnostic + Sync + Send + 'static, S: ToString>(
        err: T,
        source: S,
    ) -> String {
        let source = Arc::new(source.to_string());
        let diagnostic = Report::from(err).with_source_code(source);

        let mut reporter = Reporter::new(Style::NoColor);
        reporter.push(diagnostic);
        reporter.report_to_string()
    }

    #[test]
    fn brainfuck() {
        let text = include_str!("../../examples/brainfuck.bnfgen");
        let grammar = RawGrammar::parse(text).unwrap();
        insta::assert_debug_snapshot!(grammar);
    }

    #[test]
    fn typed() {
        let text = r#"
            <E> ::= <E: "int"> "+" <E: "int"> ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap();
        insta::assert_debug_snapshot!(grammar);
    }

    #[test]
    fn repeat() {
        let text = r#"
            <E> ::= "a" {1, 10} | "b" {2, } | "c" {3} ;
        "#;
        let grammar = RawGrammar::parse(text).unwrap();
        insta::assert_debug_snapshot!(grammar);
    }

    #[test]
    fn unexpected_eof() {
        let text = "<start> ::= \"Hello\" | \"World\""; // no semi
        let err = RawGrammar::parse(text).err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn invalid_token() {
        let text = "*";
        let err = RawGrammar::parse(text).err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn invalid_re() {
        let text = r#"<R> ::= re("["); "#;
        let err = RawGrammar::parse(text).err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn undefined_nt() {
        let text = "<E> ::= <S>;";
        let err = RawGrammar::parse(text).unwrap().to_checked().err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn duplicated_def() {
        let text = r#"
            <E> ::= <S>;
            <S> ::= <E>;
            <E> ::= "?";
        "#;
        let err = RawGrammar::parse(text).unwrap().to_checked().err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn invalid_repeat() {
        let text = r#"
            <E> ::= "a" {10, 1};
        "#;
        let err = RawGrammar::parse(text).unwrap().to_checked().err().unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn unreachable_nt() {
        let text = r#"
            <E> ::= "Hello" | <A> ;
            <W> ::= "World" ;
            <A> ::= <B> ;
            <B> ::= <A> ;
            <C> ::= <W> ;
        "#;
        let err = RawGrammar::parse(text)
            .unwrap()
            .graph()
            .check_unused("E")
            .err()
            .unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    // TODO: bug, this is also a dead loop
    // #[test]
    // fn common_dead_loop() {
    //     let text = r#"
    //         <S> ::= <S> | <S> <E> ;
    //         <E> ::= "Terminal" ;
    //     "#;
    //     let err = RawGrammar::parse(text)
    //         .unwrap()
    //         .graph()
    //         .check_trap_loop()
    //         .err()
    //         .unwrap();
    //     let ui = report_with_unnamed_source(err, text);
    //     insta::assert_snapshot!(ui);
    // }

    #[test]
    fn trap_loop() {
        let text = r#"
            <E> ::= <D> | <F>;
            <C> ::= <D> ;
            <D> ::= <C> ;
            <F> ::= <G> ;
            <G> ::= <F> | "Terminal" ;
        "#;
        let err = RawGrammar::parse(text)
            .unwrap()
            .graph()
            .check_trap_loop()
            .err()
            .unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }

    #[test]
    fn tri_loop() {
        let text = r#"
            <no-trap-01> ::= <A> | "Terminal" ;
            <no-trap-02> ::= <A> | <term> ;
            <term> ::= "Terminal" ;
            <A> ::= <B> ;
            <B> ::= <C> ;
            <C> ::= <A> ;
        "#;
        let err = RawGrammar::parse(text)
            .unwrap()
            .graph()
            .check_trap_loop()
            .err()
            .unwrap();
        let ui = report_with_unnamed_source(err, text);
        insta::assert_snapshot!(ui);
    }
}
