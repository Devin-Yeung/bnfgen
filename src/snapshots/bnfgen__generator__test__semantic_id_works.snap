---
source: src/generator.rs
expression: out
---
decl n 
 decl n 
 decl x 
 decl x 
 show n
